File: 00053_DP__Maximum_Subarray/sol2.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        if (nums.size() == 1) return nums[0];

        int _max = INT_MIN;
        _maxSubArray(nums, nums.size() - 1, _max);
        return _max;
    }
private:
    int _maxSubArray(vector<int>& nums, int i, int& _max) {
        int ans;
        if (i == 0) {
            ans = nums[i]; 
        }
        else {
            int sub = _maxSubArray(nums, i - 1, _max);
            ans = max(sub + nums[i], nums[i]);
        }
        _max = max(_max, ans);
        return ans;
    }
};

File: 00239_Array__Sliding_Window_Maximum/sol2.py

import heapq

class Solution:

    ## O(N * K)  --> time exceeded
    ## O(LogN * K)  --> ok

    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        
        h = []
        for i in range(k):
            heapq.heappush(h, (-nums[i], i))
        
        ans = [-h[0][0]]

        i = k
        while i < len(nums):
            left = i - k
            assert left >= 0

            while h and h[0][1] <= left:
                heapq.heappop(h)
            
            heapq.heappush(h, (-nums[i], i))
            ans.append(-h[0][0])
            i += 1
        
        return ans

            

